### DISCUSSION TITLE
3D spinodal decomposition

### DISCUSSION THREAD

**Discussion by Amir1361** [idaholab/moose#17427] [Q&A Modules: Phase field] [answered]
Hello,

I am trying to have a 3D simulation of spinodal decomposition. First, I used "GeneratedMesh," and it works with a small mesh number. But when I increased the nx, ny, and nz, I was out of memory. 

"mpiexec noticed that process rank 0 with PID 66797 on node cpu110 exited on signal 9 (Killed)."

Then, I changed it to "DistributedRectilinearMeshGenerator" to run it with more mesh numbers. But I got the following error:

"^[[31m
*** ERROR ***
Task init_mesh is not registered to build MeshGenerator derived objects^[[39m

--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
[cpu102:07986] 7 more processes have sent help message help-mpi-api.txt / mpi-abort
[cpu102:07986] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages"

Here is my mesh part in the input file:

[Mesh]
  type = DistributedRectilinearMeshGenerator
  dim = 3
  elem_type = QUAD4
  nx = 25
  ny = 25
  nz = 25
  xmax = 100
  ymax = 100
  zmax = 100
[]

Could you please help me to solve this?

Best,
Amir

**Comment by aeslaughter** [comment, marked as answer]
Do you have mesh adaptivity enabled?

**Reply by jiangwen84** [reply]
On top of @aeslaughter 's question, are you using a direct solver, i.e. LU? 

BTW, your element_type = QUAD4 does not make sense for a 3D domain.

**Reply by Amir1361** [reply]
Thank you for your replies. 
This is my input file: [FeCrCo_3D.txt](https://github.com/idaholab/moose/files/6202370/FeCrCo_3D.txt)

I removed the "element_type = QUAD4", but it doesn't work yet. How can I enable the mesh adaptivity?

**Reply by permcody** [reply]
Even with `DistributedRectilinearMeshGenerator` you still need to set your `parallel_type = distributed` to run with distributed mesh. However, I don't think this problem is big enough to require distributed so I tried it without. It ran OK as the per core memory usage was about just over 2GB after 3 time steps. Depending on your system and how many MPI ranks you used, you can figure out if that would exceed your available memory. I would expect that number to continue to climb, but not too much beyond that. I then tried it with distributed and the memory usage stayed well under 2GB (~1.3GB).

Another very interesting thing I noted was how slow `DistributedRectilinearMeshGenerator` was. It took a very long time to generate the mesh. Using the simple "GeneratoredMesh" object it was only a split second. I know the former uses less memory but I'm a little surprised with how much slower it was. We probably haven't tried it much with medium size problems like this. It really is for very large problems. 

So your input file works, just make sure you aren't running yourself out of memory!

**Reply by jiangwen84** [reply]
1. Adding a sub-block will fix the error.
[Mesh]
  [drmg]
    type = DistributedRectilinearMeshGenerator
    dim = 3
    nx = 25
    ny = 25
    nz = 25
    xmax = 100
    ymax = 100
    zmax = 100
  []
[]

2. I am not sure if switching to distributed mesh will resolve your memory issue. I ran your input file with 10 MPI procs, and roughly each process requires a maximum 3-4 Gb memory. That seems to be reasonable to me. You might want to request more memory on you HPC to large 3D simulation.

**Reply by Amir1361** [reply]
Thank you very much! I do not have any problem with following mesh size:

[Mesh]
[drmg]
type = DistributedRectilinearMeshGenerator
dim = 3
nx = 25
ny = 25
nz = 25
xmax = 100
ymax = 100
zmax = 100
[]
[]

But when I want to increase nx, ny, nz and xmax, ymax, zmax:

[Mesh]
[drmg]
type = DistributedRectilinearMeshGenerator
dim = 3
nx = 100
ny = 100
nz = 100
xmax = 200
ymax = 200
zmax = 200
[]

I get the memory error:

"mpiexec noticed that process rank 0 with PID 142272 on node cpu102 exited on signal 9 (Killed)."

I am running my job on cluster computer (Boise State University). Is there any way to solve this issue? Thank you very much for your help!

Amir

**Reply by jiangwen84** [reply]
To run a 3D problem with 100x100x100 definitely requires a lot of memory usage:-) So you have to work with your HPC admin to increase the memory for your simulation.

**Reply by permcody** [reply]
Amir, I suggest that when you run, you start up a "top" process on the lead node in your batch job and watch memory usage to see where you are at. If you are only a little over (it runs for past the mesh stage and dies in the solver), one way to get the extra memory you need is to simply not use all the cores on each node (e.g. if your nodes have 32 processors each, try running your problem using only 24 or 28 cores on the nodes). Doing this with a batch scheduling system is fairly straightforward. Your million element mesh shouldn't take too much memory (Several hundred MB on each rank), so there are other factors that may be causing you to run out of memory related to the sizes of your equation systems or other resident memory. Let us know if you need any further guidance.

**Reply by Amir1361** [reply]
@permcody Thank you very much for your comments. I will test the different options. Thank you for your help; I appreciate it.
